### PR TITLE
Context Window Exceeded fix

### PR DESCRIPTION
**Pull request by enyst**
**End-user friendly description of the problem this fixes or functionality that this introduces**

- [x] Include this change in the Release Notes. If checked, you must provide an **end-user friendly** description for your change below
Truncate agent history when context window is exceeded

---
**Give a summary of what the PR does, explaining any non-trivial design decisions**

This PR proposes a simple fix for the ContextWindowExceededError:
- when raised, truncate the context roughly in half
- save state and reload the truncated history when session is restored

Please note that this is not well tested, but it is ready for code review.

---
**Link of any specific issues this addresses**
Fix https://github.com/All-Hands-AI/OpenHands/issues/4951
Fix https://github.com/All-Hands-AI/OpenHands/issues/4894

---

To run this PR locally, use the following command:
```
docker run -it --rm   -p 3000:3000   -v /var/run/docker.sock:/var/run/docker.sock   --add-host host.docker.internal:host-gateway   -e SANDBOX_RUNTIME_CONTAINER_IMAGE=docker.all-hands.dev/all-hands-ai/runtime:2fcb088-nikolaik   --name openhands-app-2fcb088   docker.all-hands.dev/all-hands-ai/openhands:2fcb088
```

### CODE REVIEW

File: openhands/controller/agent_controller.py [hunk 1]
```diff
@@ -680,16 +696,51 @@ def _init_history(self):
...
-                start_id=start_id,
-                end_id=end_id,
-                reverse=False,
-                filter_out_type=self.filter_out,
-                filter_hidden=True,
+        events: list[Event] = []
+
+        # If we have a truncation point, get first user message and then rest of history
+        if hasattr(self.state, 'truncation_id') and self.state.truncation_id:
+            # Get events from start_id until we find first user message
```
Comment (enyst) [inline openhands/controller/agent_controller.py:703]:
```suggestion
```

File: openhands/controller/agent_controller.py [hunk 2]
```diff
@@ -680,16 +696,50 @@ def _init_history(self):
...
+                    filter_out_type=self.filter_out,
+                    filter_hidden=True,
+                )
+            )
+            events.extend(truncated_events)
+        else:
+            # Normal loading from start_id
+            events = list(
+                self.event_stream.get_events(
+                    start_id=self.state.start_id if self.state.start_id >= 0 else 0,
```
Comment (enyst) [inline openhands/controller/agent_controller.py:736]:
```suggestion
                    start_id=start_id,
```

**Review by neubig — APPROVED** [review]
This looks good to me, thank you for doing it!

It'd be nice if we could do an eval before merging just to make sure that things are working, I added the label now.

**Comment by enyst** [comment]
Unfortunately it looks like the [Deepseek eval workflow](https://github.com/All-Hands-AI/OpenHands/actions/runs/11789042557) doesn't work on `main`. 🤔 I seem to recall that we had merged it with high hopes, but function calling didn't actually work with Deepseek, and this workflow was not really used yet afaik.

I can test tomorrow.

**Comment by neubig** [comment]
I think we don't need to test then, this should be fine!

**Pull request merged**